File: examples/python/rps_client/server_manager.py
"""Manages the rps-server subprocess lifecycle."""

import os
import signal
import socket
import subprocess
import sys
import time
from typing import TYPE_CHECKING
from pathlib import Path

import grpc

if TYPE_CHECKING:
    from rps_client.windows_job import WindowsJobObject


class ServerManager:
    """Spawns and kills the rps-server binary as a subprocess."""

    def __init__(
        self,
        server_bin: str = "rps-server",
        port: int = 50051,
        db: str = "rps-plugins.db",
    ):
        self.server_bin = server_bin
        self.port = port
        self.db = db
        self._process: subprocess.Popen | None = None
        self._windows_job: WindowsJobObject | None = None
        self._old_sigint_handler = None
        self._old_sigterm_handler = None
        self._debug = os.getenv("RPS_DEBUG_PROCESS_LIFECYCLE") == "1"

    def _handle_signal(self, signum, frame):
        """Signal handler to ensure cleanup on SIGINT/SIGTERM."""
        self.stop(in_hurry=True)
        if signum == signal.SIGINT and self._old_sigint_handler:
            self._old_sigint_handler(signum, frame)
        elif signum == signal.SIGTERM and self._old_sigterm_handler:
            self._old_sigterm_handler(signum, frame)
        else:
            sys.exit(1)

    def start(self, timeout: float = 10.0) -> None:
        """Start the server subprocess and wait for it to accept connections."""
        server_path = Path(self.server_bin)
        if not server_path.exists() or not server_path.is_file():
            raise FileNotFoundError(f"Cannot find rps-server binary at: {self.server_bin}")

        scanner_name = "rps-pluginscanner.exe" if sys.platform == "win32" else "rps-pluginscanner"
        scanner_path = server_path.with_name(scanner_name)
        if not scanner_path.exists() or not scanner_path.is_file():
            raise FileNotFoundError(f"Cannot find {scanner_name} alongside rps-server at: {scanner_path}")

        self._old_sigint_handler = signal.getsignal(signal.SIGINT)
        self._old_sigterm_handler = signal.getsignal(signal.SIGTERM)
        signal.signal(signal.SIGINT, self._handle_signal)
        signal.signal(signal.SIGTERM, self._handle_signal)

        cmd = [
            self.server_bin,
            "--port", str(self.port),
            "--db", self.db,
        ]
        self._process = subprocess.Popen(
            cmd,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            start_new_session=(sys.platform != "win32"),
        )
        if self._debug:
            print(f"[rps] spawned server pid {self._process.pid}")

        if sys.platform == "win32":
            # Windows process ownership model:
            # parent/child does not imply lifetime ownership. Attach rps-server
            # to a Job Object so parent exit tears down the process tree.
            from rps_client.windows_job import WindowsJobObject

            try:
                self._windows_job = WindowsJobObject.create_and_assign(self._process.pid)
                if self._debug:
                    print(f"[rps] attached Windows Job Object to pid {self._process.pid}")
            except Exception:
                self._process.terminate()
                self._process.wait(timeout=2.0)
                self._process = None
                raise

        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline:
            if self._process.poll() is not None:
                self._close_windows_job()
                raise RuntimeError(
                    f"rps-server exited immediately with code {self._process.returncode}"
                )
            try:
                with socket.create_connection(("localhost", self.port), timeout=0.5):
                    return
            except (ConnectionRefusedError, OSError, socket.timeout):
                time.sleep(0.2)

        self.stop(in_hurry=True)
        raise TimeoutError(
            f"rps-server did not start within {timeout}s on port {self.port}"
        )

    def stop(self, in_hurry: bool = False) -> None:
        """Stop the server subprocess gracefully, then forcefully if needed."""
        if self._old_sigint_handler:
            signal.signal(signal.SIGINT, self._old_sigint_handler)
            self._old_sigint_handler = None
        if self._old_sigterm_handler:
            signal.signal(signal.SIGTERM, self._old_sigterm_handler)
            self._old_sigterm_handler = None

        if self._process is None:
            self._close_windows_job()
            return

        try:
            if in_hurry:
                if self._debug:
                    print("[rps] stop(in_hurry=True)")
                self._process.terminate()
                try:
                    self._process.wait(timeout=3.0)
                except subprocess.TimeoutExpired:
                    self._process.kill()
                    self._process.wait()
                return

            try:
                from rps_client.proto import rps_pb2, rps_pb2_grpc

                channel = grpc.insecure_channel(f"localhost:{self.port}")
                stub = rps_pb2_grpc.RpsServiceStub(channel)
                stub.Shutdown(rps_pb2.ShutdownRequest(), timeout=2)
                channel.close()
            except Exception:
                pass

            try:
                self._process.wait(timeout=5.0)
            except subprocess.TimeoutExpired:
                self._process.terminate()
                try:
                    self._process.wait(timeout=2)
                except subprocess.TimeoutExpired:
                    self._process.kill()
                    self._process.wait()
        finally:
            if self._debug:
                print("[rps] server process stopped")
            self._process = None
            self._close_windows_job()

    def _close_windows_job(self) -> None:
        if self._windows_job is not None:
            try:
                self._windows_job.close()
                if self._debug:
                    print("[rps] closed Windows Job Object")
            finally:
                self._windows_job = None

    @property
    def address(self) -> str:
        return f"localhost:{self.port}"

    @property
    def is_running(self) -> bool:
        return self._process is not None and self._process.poll() is None

    def __enter__(self):
        self.start()
        return self

    def __exit__(self, *exc):
        self.stop()

File: examples/java/src/main/java/rps/client/ServerManager.java
package rps.client;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

public class ServerManager implements AutoCloseable {
    private static final boolean PROCESS_DEBUG =
            "1".equals(System.getenv("RPS_DEBUG_PROCESS_LIFECYCLE"));

    private final String binPath;
    private final int port;
    private final String dbPath;
    private final String logLevel;
    private Process process;
    private WindowsJobObject windowsJob;

    public ServerManager(String binPath, int port, String dbPath, String logLevel) {
        this.binPath = binPath;
        this.port = port;
        this.dbPath = dbPath;
        this.logLevel = logLevel;
    }

    public void start() throws IOException, InterruptedException {
        File serverFile = new File(binPath);
        if (!serverFile.exists() || !serverFile.isFile()) {
            throw new FileNotFoundException("Cannot find rps-server binary at: " + binPath);
        }

        String os = System.getProperty("os.name").toLowerCase();
        String scannerName = os.contains("win") ? "rps-pluginscanner.exe" : "rps-pluginscanner";
        File scannerFile = new File(serverFile.getParentFile(), scannerName);
        if (!scannerFile.exists() || !scannerFile.isFile()) {
            throw new FileNotFoundException("Cannot find " + scannerName + " alongside rps-server at: " + scannerFile.getAbsolutePath());
        }

        List<String> cmd = new ArrayList<>();
        cmd.add(binPath);
        cmd.add("--port");
        cmd.add(String.valueOf(port));
        cmd.add("--db");
        cmd.add(dbPath);
        cmd.add("--log-level");
        cmd.add(logLevel);

        ProcessBuilder pb = new ProcessBuilder(cmd);
        pb.redirectError(ProcessBuilder.Redirect.DISCARD);
        pb.redirectOutput(ProcessBuilder.Redirect.DISCARD);

        System.out.println("Spawning server: " + String.join(" ", cmd));
        this.process = pb.start();
        if (isWindows()) {
            // Windows process ownership model:
            // parent/child does not imply lifetime ownership. We explicitly attach
            // rps-server to a Job Object so parent exit tears down the process tree.
            this.windowsJob = WindowsJobObject.createAndAssign(process.pid());
            if (PROCESS_DEBUG) {
                System.out.println("[rps] attached Windows Job Object to pid " + process.pid());
            }
        }

        // Wait for server to be ready
        long deadline = System.currentTimeMillis() + 10000;
        while (System.currentTimeMillis() < deadline) {
            if (!process.isAlive()) {
                throw new IOException("rps-server exited immediately with code " + process.exitValue() + 
                                     ". Check rps-server.log for details.");
            }
            try (Socket socket = new Socket()) {
                socket.connect(new InetSocketAddress("127.0.0.1", port), 500);
                return; // Connected!
            } catch (IOException e) {
                Thread.sleep(200);
            }
        }
        throw new IOException("rps-server did not start within 10s on port " + port);
    }

    public static String findServerBinary() {
        // Only check CWD and the directory of the running class/JAR
        String[] names = {"rps-server", "rps-server.exe"};
        String[] locations = {
            System.getProperty("user.dir"),
            new File(ServerManager.class.getProtectionDomain().getCodeSource().getLocation().getPath()).getParent()
        };

        for (String loc : locations) {
            if (loc == null) continue;
            for (String name : names) {
                File f = new File(loc, name);
                if (f.exists() && f.canExecute() && !f.isDirectory()) {
                    return f.getAbsolutePath();
                }
            }
        }

        // Try PATH
        String pathEnv = System.getenv("PATH");
        if (pathEnv != null) {
            String sep = System.getProperty("path.separator");
            for (String p : pathEnv.split(sep)) {
                for (String name : names) {
                    File f = new File(p, name);
                    if (f.exists() && f.canExecute() && !f.isDirectory()) {
                        return f.getAbsolutePath();
                    }
                }
            }
        }

        return null;
    }

    @Override
    public void close() {
        try {
            if (process != null) {
                if (PROCESS_DEBUG) {
                    System.out.println("[rps] stopping server pid " + process.pid());
                }
                process.destroy();
                try {
                    process.waitFor(2, TimeUnit.SECONDS);
                } catch (InterruptedException ignored) {
                }

                // If graceful stop didn't finish quickly, force-kill child tree.
                if (process.isAlive()) {
                    process.descendants().forEach(ProcessHandle::destroyForcibly);
                    process.destroyForcibly();
                    try {
                        process.waitFor(5, TimeUnit.SECONDS);
                    } catch (InterruptedException ignored) {
                    }
                }
            }
        } finally {
            if (windowsJob != null) {
                windowsJob.close();
                windowsJob = null;
                if (PROCESS_DEBUG) {
                    System.out.println("[rps] closed Windows Job Object");
                }
            }
            process = null;
        }
    }

    private static boolean isWindows() {
        return System.getProperty("os.name").toLowerCase().contains("win");
    }
}
